Implementation of uTVM

weberlo opened the pull request:
This PR provides an initial implementation of the ideas discussed in #2563.  It has been a joint effort between @Mutinifni, @tqchen, and me to design and implement this system.

We only include an implementation of the emulated host low-level device interface in this PR.  The OpenOCD interface will be included in a future PR.

Currently, the emulated host device can run ResNet18.  As the implementation matures, we will try more complex models.

CC @huajsj @yongwww @jroesch @Ravenwater @junrushao1994 @joshpoll

weberlo commented on src/api/api_pass.cc:
```diff
@@ -130,6 +130,7 @@ REGISTER_PASS(RewriteUnsafeSelect);
 REGISTER_PASS(Inline);
 REGISTER_PASS(IRTransform);
 REGISTER_PASS(VectorizeLoop);
+REGISTER_PASS(SkipVectorize);
```
I missed adding this line in https://github.com/dmlc/tvm/pull/3222, so I added it here.

tqchen commented on tests/python/unittest/test_runtime_micro.py:
```diff
@@ -0,0 +1,183 @@
+    assert result.sum() != 0.0
+
+
+def test_resnet_pretrained():
```
Let us remove the end to end test for now

tqchen replied:
please update per comment

weberlo replied:
@tqchen Is adding the `@nottest` decorator sufficient?  If not, I'd still like to have this code somewhere, even if on a branch on a private repo.

tqchen commented on topi/python/topi/generic/nn.py:
```diff
@@ -24,8 +24,9 @@ def _default_schedule(outs, auto_inline):
     """Default schedule for llvm."""
     target = tvm.target.current_target(allow_none=False)
     outs = [outs] if isinstance(outs, tvm.tensor.Tensor) else outs
-    if target.target_name != "llvm":
```
if target.target_name not in ("llvm", "c")

tqchen commented on python/tvm/contrib/binutil.py:
```diff
@@ -0,0 +1,218 @@
+    """
+    tmp_dir = util.tempdir()
+    rel_obj = tmp_dir.relpath("relocated.o")
+    ld_script_contents = '''
```
try to be consistent and use """ instead of '''

tqchen commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,125 @@
+    _MicroInit(device_type, runtime_lib_path, port)
+
+
+def from_host_mod(host_mod, device_type):
```
from_source_module

tqchen commented on python/tvm/micro/cross_compile.py:
```diff
@@ -0,0 +1,48 @@
+from .._ffi.base import py_str
+
+
+def create_lib(output, sources, options=None, cc="gcc"):
```
consider patch contrib.cc to introduce a cross compiler https://github.com/dmlc/tvm/blob/master/python/tvm/contrib/cc.py#L90 to move the fPIC and -shared to the cross_compiler part, so we can reuse these functions instead of creating our own.

tqchen replied:
pending consolidation to cc

weberlo replied:
I thought we were doing that in a separate PR?

tqchen replied:
OK. that also works

tqchen commented on src/runtime/micro/device/utvm_runtime.h:
```diff
@@ -0,0 +1,73 @@
+ * certain backends such as OpenGL.
+ * \return nullptr when error is thrown, a valid ptr if success
+ */
+void* TVMBackendAllocWorkspace(int device_type, int device_id, uint64_t size,
```
nit, one argument per line

tqchen commented on src/runtime/micro/low_level_device.h:
```diff
@@ -0,0 +1,77 @@
+   * \brief getter function for base_addr
+   * \return the base address of the device memory region
+   */
+  virtual dev_base_addr base_addr() const = 0;
```
consider using CamelCase DeviceBaseAddr

tqchen commented on src/runtime/micro/micro_common.h:
```diff
@@ -0,0 +1,272 @@
+};
+
+// TODO(weberlo): There's a lot of duplication between these classes.  How can we consolidate?
+class dev_addr;
```
CamelCase

tqchen commented on src/runtime/micro/micro_common.h:
```diff
@@ -0,0 +1,272 @@
+
+  std::uintptr_t val() const { return val_; }
+  template <typename T>
+  T* as_ptr() const { return reinterpret_cast<T*>(val_); }
```
operator==(nullptr_t)

tqchen commented on src/runtime/micro/micro_common.h:
```diff
@@ -0,0 +1,272 @@
+class dev_base_offset;
+
+/*! \brief absolute device address */
+class dev_addr {
```
Consider using a base class with https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern to reduce the implementation duplication.

OpaquePtr

tqchen commented on src/runtime/micro/micro_common.h:
```diff
@@ -0,0 +1,272 @@
+
+  std::uintptr_t val() const { return val_; }
+  template <typename T>
+  T* as_ptr() const { return reinterpret_cast<T*>(val_); }
```
document all the API

tqchen commented on src/runtime/micro/micro_common.h:
```diff
@@ -0,0 +1,272 @@
+  explicit dev_addr(std::nullptr_t) : val_(0) {}
+  ~dev_addr() {}
+
+  std::uintptr_t val() const { return val_; }
```
value()

tqchen commented on src/runtime/micro/micro_common.h:
```diff
@@ -0,0 +1,272 @@
+ * \param align alignment of the returned size (default: 8)
+ * \return size of the section if it exists, 0 otherwise
+ */
+size_t GetSectionSize(std::string binary_name, SectionKind section,
```
one argument per line

tqchen commented on src/runtime/micro/target_data_layout_encoder.h:
```diff
@@ -0,0 +1,163 @@
+};
+
+template <typename T>
+TargetDataLayoutEncoder::Slot<T>::Slot(TargetDataLayoutEncoder* parent, size_t start_offset,
```
one argument per line

tqchen commented:
cc @mjs-arm @kparzysz

tqchen commented on include/tvm/runtime/micro/utvm_device_lib.h:
```diff
@@ -0,0 +1,61 @@
+void (*TVMAPISetLastError_)(const char*) = (void (*)(const char*)) 1;
+
+#ifdef __cplusplus
+extern "C"
```
you only need to put extern C once, in the begin and end, 


```c
#ifdef __cplusplus
extern "C" {
#endif 

// your code 

#ifdef __cplusplus
}
#endif
```

tqchen commented on src/runtime/micro/device/utvm_runtime.c:
```diff
@@ -0,0 +1,46 @@
+// `NULL`.  Why?
+
+// These pointers are patched at load time to point to the workspace section.
+char *workspace_start = (char*) 1;
```
use meaningful name utvm_workspace_begin

tqchen commented on src/runtime/micro/device/utvm_runtime.h:
```diff
@@ -0,0 +1,73 @@
+ * \brief POD variant of TVMArgs
+ */
+typedef struct {
+  TVMValue* values;
```
document each field

tqchen commented on src/runtime/micro/device/utvm_runtime.h:
```diff
@@ -0,0 +1,73 @@
+  UTVMArgs* args;
+} UTVMTask;
+
+// TODO(weberlo): Remove duplicate docs?
```
it is OK to keep duplicated docs as long as a user have a good context

tqchen commented on src/runtime/micro/micro_session.cc:
```diff
@@ -0,0 +1,309 @@
+        tvm_vals_slot.Write(&val_addr);
+        break;
+      }
+      // TODO(mutinifni): implement other cases if needed
```
need to add implementation that passes in double, and int64

weberlo replied:
I'd like to add a test for this.  Is there an example program that generates scalar values?

weberlo replied:
^-- this question

weberlo replied:
@tqchen

tqchen commented on src/runtime/micro/micro_session.cc:
```diff
@@ -0,0 +1,309 @@
+    switch (type_codes[i]) {
+      case kNDArrayContainer:
+      case kArrayHandle: {
+        TVMArray* arr_handle = reinterpret_cast<TVMArray*>(args->values[i].v_handle);
```
don't do an explicit cast. directly do TVMArray* arr_handle = args->values[i]; this will invoke the conversion function. To help us to be future compatible.

tqchen commented on src/runtime/micro/micro_session.cc:
```diff
@@ -0,0 +1,309 @@
+  const int* type_codes = args->type_codes;
+  int num_args = args->num_args;
+
+  auto tvm_vals_slot = encoder->Alloc<TVMValue*>(num_args);
```
why allocate TVMValue's pointer instead of the value itself?

tqchen commented on src/runtime/micro/micro_session.cc:
```diff
@@ -0,0 +1,309 @@
+      case kNDArrayContainer:
+      case kArrayHandle: {
+        TVMArray* arr_handle = reinterpret_cast<TVMArray*>(args->values[i].v_handle);
+        TVMValue* val_addr = EncoderWrite(encoder, arr_handle).as_ptr<TVMValue>();
```
void* arr_ptr = EncoderWrite(encoder, arr_handle).cast_to<void*>();
TVMValue val;
val.v_handle = arr_ptr;
tvm_vals_slot.Write(val);

weberlo commented:
I haven't addressed all of the feedback yet.  @tqchen suggested I replace the `DevAddr`, `DevBaseAddr`, and `DevBaseOffset` classes with a single `DevicePtr` class for addresses, and then we would use `size_t` for offsets.  I still need to think some more on whether this is the way to go.

liangfu commented on src/runtime/micro/micro_common.h:
```diff
@@ -0,0 +1,279 @@
+ * \brief enum of device memory region sections
+ */
+enum SectionKind : int {
+  kText = 0,
```
Do we need to add prefix to these constants?

weberlo replied:
The leading "k" seems to be the convention, unfortunately.

liangfu replied:
I mean we may need an additional prefix between the leading `k` and `Text`, like the definition of `kDLFloat` in the dlpack. Without such prefix, a developer might confuse between `SectionKind` and other enums, because `kText` and `kString` seems to belong to a same enum.

weberlo replied:
I don't have any strong feelings about adding a prefix like `Micro` to them, but I'm unsure of the conventions around enums.  CC @tqchen

liangfu replied:
I just have a feeling that someone might confuse some of the constants in this enum with `kStr` and `kBytes` in `TVMTypeCode` enum.

tqchen replied:
Let us use enum class, which will enforce the prefix

weberlo commented on src/runtime/micro/device/utvm_runtime.c:
```diff
@@ -0,0 +1,45 @@
+}
+
+int TVMBackendFreeWorkspace(int device_type, int device_id, void* ptr) {
+  // TODO(weberlo): Actually free memory.
```
This is something I wanted to discuss.  To me, it's unclear how we should go about freeing memory on the device, since we don't have access to the standard library.

One potentially easy solution is to continue ignoring calls to this function, but reset the workspace pointers between each function call (i.e., each `UTVMTask`).

Maybe @tqchen has some thoughts?

tqchen replied:
Our original proposal was to keep a counter of number of units being allocated, decrease the pointer(but not freeing the memory). and free everything once the number goes to zero

weberlo replied:
Oh right.  Forgot about that.

weberlo commented on include/tvm/runtime/micro/utvm_device_lib.h:
```diff
@@ -0,0 +1,52 @@
+  (*TVMAPISetLastError_)(msg);
+}
+
+float min(float a, float b) {
```
We can't rename these functions without modifying the C codegen, because `min` and `max` are the names that are generated by the source that includes this header.

As for switching to macros, I would think modern compilers are smart enough to inline small functions such as this.  I could be wrong though.

liangfu replied:
For modern compilers, such functions are defined within certain namespace like `std`, and they typically use `template` based implementation. 

However, this header file is more likely to be a C header. Therefore, following the C convention is more likely to be reasonable. Otherwise, someone who wanted to use `double` might lose precision unintentionally.

liangfu replied:
I think it would be fine to add `#include <math.h>` in `c_runtime_api.h`, and rename `min` to `fmin` in `codegen_c.cc`, and avoid implementing a custom compare function.

weberlo replied:
Adding an include for those functions won't help here, because we don't currently link against libc, so we don't have implementations for those functions.  We have considered loading something like [openlibm](https://github.com/JuliaMath/openlibm) on the device, to get access to math functions, but that would likely be in a future PR.

liangfu replied:
Okay, let's put this temp implement of min/max functions right here.

However, I'm feeling that integration with something like `openlibm` wouldn't be the ultimate goal of TVM. Instead, these functions should be built in TVM, either in the `runtime`, in `topi` or `codegen`. (Actually, they are already been implemented in the later two options.)

weberlo replied:
Yeah.  If there already exist implementations of some of these functions, it might make more sense to modify codegen to inline those implementations in the generated source file.  We'll see what makes sense as we spend more time using MicroTVM.

tqchen replied:
Let us simply modify the CodegenC to directly emit conditional expression ```((x)>(y)? (x):(y))```

weberlo replied:
Will any of the generated expressions contain mutation?  If so, we'll need to create bindings for `x` and `y` to prevent duplicate computations.  If there's no mutation, we can safely inline, but it's still inefficient.

weberlo replied:
@tqchen As far as I understand, there _are_ some operators that are stateful, so while the current codegen modification works for ResNet, I don't know if we want to merge this change in.

u99127 replied:
@liangfu - On the topic about what comes from libm vs what comes from the micro tvm device run time we need to think about where to draw the line with what comes in the C runtimes from static toolchains that we would be using to compile this. For instance if this is to be using a variant of a baremetal gnu toolchain as has been suggested (to compile the generated C backend file), the library of choice these days for baremetal stuff is actually newlib (and newlib-nano) and thus you get some of the libm functions for free in this regard. Putting in all traditional math runtime functions in  a micro tvm runtime sounds like the wrong call to me as you'd need to get this anyway from the static toolchain that was being used to link things in.

On the topic of macros vs inlines, if you put these out as static always_inline functions you get the same effect, inlining works also at -O0. In fact this is a common technique used in various intrinsic header files that are provided by GCC to force the inlining for intrinsics, thus you could avoid macros.

I hope this helps.

liangfu replied:
@u99127 It comes more clear to me that C backend is very much different from LLVM backend, which generates instruction-level optimized code directly. And I agree that integrating with either `newlib` or `newlib-nano` should be a preferable choice to complement the baremetal C runtime. However, IMHO, the C codegen should be designed for C (with some compatibility with C++), not exclusively for C++. For instance, we should avoid using `max` for both `float` and `double`, instead, `fmaxf` and `fmax` should be used (, even in this temp implement of the C math library).

On the macros vs inlines topic, I agree that inlines is preferred over macros. For instance, `MIN(MAX(a, b), c)` can be simply expanded into `(a>b?a:b)<c?(a>b?a:b):c`, therefore, the comparison between `a` and `b` could have been performed twice.

tqchen replied:
I still think that we should directly implement it as a conditional expression in the C codegen. At the moment there is no problem of stateful expressions. If we do in the future, they will be separated into a different statement, which would not affect the semantics.

To be more careful in the future, we can just print a new variable assignment for each sub expression. and then refer to the variable. see https://github.com/dmlc/tvm/blob/cffb4fba03ea582417e2630bd163bca773756af6/src/codegen/codegen_c.cc#L129

tqchen commented on CMakeLists.txt:
```diff
@@ -36,6 +36,7 @@ tvm_option(USE_RELAY_DEBUG "Building Relay in debug mode..." OFF)
 tvm_option(USE_SGX "Build with SGX" OFF)
 tvm_option(USE_RTTI "Build with RTTI" ON)
 tvm_option(USE_MSVC_MT "Build with MT" OFF)
+tvm_option(USE_MICRO "Build with Micro" OFF)
```
Micro -> uTVM

tqchen commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,128 @@
+    # ".o" suffixes with ".obj".
+    if obj_path.endswith(".o"):
+        # TODO(weberlo): Use TVM Python logging mechanism, if there is one.
+        print("WARNING: create_micro_lib: \".o\" suffix in \"{}\" has been replaced with \".obj\"")
```
logging.warn

tqchen commented on src/runtime/micro/micro_session.cc:
```diff
@@ -0,0 +1,306 @@
+  init_binary_path_ = path;
+}
+
+DevAddr MicroSession::EncoderAppend(TargetDataLayoutEncoder* encoder, TVMArgs& args) {
```
use const TVMArgs& args

tqchen commented on src/runtime/micro/low_level_device.h:
```diff
@@ -0,0 +1,77 @@
+#ifndef TVM_RUNTIME_MICRO_LOW_LEVEL_DEVICE_H_
+#define TVM_RUNTIME_MICRO_LOW_LEVEL_DEVICE_H_
+
+#include <cstddef>
```
check if we need cstddef

tqchen commented on src/runtime/micro/micro_session.cc:
```diff
@@ -0,0 +1,306 @@
+        break;
+    }
+  }
+  type_codes_slot.WriteRaw(type_codes, num_args);
```
Consider add another API WriteArray, so it can handle the endianness issue.

weberlo replied:
It seems like `WriteRaw` is already functioning as a `WriteArray` method, and it should probably be renamed to that.

To deal with endianness issues, I think we'll need to change the way we write values to the encoder.  In particular, we'll need to store the host and target endianness in the encoder and do conversions when they don't match.  Then we'll need to disallow *raw* writes and instead implement write methods for every compound and primitive data type that we want to write to the device.  This design would essentially be traits.  I think we should make this change in a future PR though, once we have more experience with different devices (e.g., RISC-V).

liangfu replied:
TVM already provides byte swap routines in binay file handlers, I think the only effort required is to detect endianness on device, and provide the host such info.

tqchen replied:
Let us rename to WriteArray for now. The difference is WriteRaw just performs memcpy, while WriteArray will detect endianness and do endian swapping

tqchen commented on src/runtime/micro/micro_session.cc:
```diff
@@ -0,0 +1,306 @@
+  return utvm_args_slot.start_addr();
+}
+
+DevAddr MicroSession::EncoderAppend(TargetDataLayoutEncoder* encoder, TVMArray& arr) {
```
const TVMArray&

tqchen replied:
always pass by const reference if you do not need to mutate the content

tqchen commented on src/runtime/micro/target_data_layout_encoder.h:
```diff
@@ -0,0 +1,175 @@
+
+#include <algorithm>
+#include <cstring>
+#include <iomanip>
```
Check the headers and see if we really need them (put them in cc files which use them instead of the header). We want to minimize the header dependencies

weberlo replied:
will do, but do we not have a linter that checks for this?

tqchen replied:
Unfortunately we do not

tqchen commented on tests/python/unittest/test_runtime_micro.py:
```diff
@@ -0,0 +1,204 @@
+    test_workspace_add()
+    test_graph_runtime()
+    test_resnet_random()
+    # TODO(weberlo): Uncomment this test (or add it as a tutorial?)
```
we have to remove the code itself, the CI (nose) will test all functions with test_ prefix

weberlo replied:
Adding a `@nottest` annotation should work, right?

tqchen commented on src/runtime/micro/micro_module.cc:
```diff
@@ -0,0 +1,146 @@
+   * \param args type-erased arguments passed to the function
+   */
+  void RunFunction(std::string func_name, DevBaseOffset func_offset, TVMArgs args) {
+    // TODO(weberlo): Why do we need `func_name`?
```
Function name is not necessary

tqchen commented on src/runtime/micro/micro_module.cc:
```diff
@@ -0,0 +1,146 @@
+   * \brief initializes module by establishing device connection and loads binary
+   * \param binary_path path of the binary to be loaded
+   */
+  void InitMicroModule(const std::string binary_path) {
```
always use const reference(const std::string&) when you don't need to modify the content

tqchen commented:
@weberlo please fix the CI error and make sure CI is green. @liangfu please help to do another round of review. also cc @grwlf @Mutinifni @MarisaKirisame @kazum @vinx13. Please help review if you can.

Mutinifni commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,142 @@
+    compile_cmd : str, optional
+        compiler command to be used
+
+    obj_path : str, optional
```
`obj_path` both in function parameters and return, might be confusing.

Mutinifni commented on src/codegen/codegen_c_host.cc:
```diff
@@ -37,6 +37,7 @@ CodeGenCHost::CodeGenCHost() {
 void CodeGenCHost::Init(bool output_ssa) {
   decl_stream << "#include \"tvm/runtime/c_runtime_api.h\"\n";
   decl_stream << "#include \"tvm/runtime/c_backend_api.h\"\n";
+  decl_stream << "#include \"tvm/runtime/micro/utvm_device_lib.h\"\n";
```
It might make sense to make this include optional based on whether uTVM is being used. The C codegen could be used for other purposes too (e.g., running directly on the host).

weberlo replied:
Yeah.  This is one thing I wasn't sure about, because we also need to modify `codegen_c.cc` (for the ternary operator changes suggested by Tianqi).  So do we want `micro_tvm` flags in both files?

Mutinifni replied:
@tqchen would appreciate your suggestion here. I'm not sure if we want to involve runtime-level info in the codegen.

tqchen replied:
If we disable ternary operators(to support min/max), we do not need to add any additional declaration headers for now.  

In the future, for other extern functions that needs to be declared. We should have a way to automatically extract a function signature from the call arguments, and write the declaration to decl_stream, without having to include additional header files.

tqchen replied:
The source file being compiled don't need to contain the include, instead, during build, we use the command

```c
gcc -o test_lib.o  micro_tvm_generated.cc src/runtime/micro/utvm_device_lib.cc
```
Where  ```src/runtime/micro/utvm_device_lib.cc``` contains the definition of these functions

weberlo replied:
To be clear, you're suggesting we duplicate the definitions for every created library?

weberlo replied:
It looks like this would pose a problem with checking device errors, because `MicroSession` only has access to the symbol table for the µTVM runtime.  Currently, we just read from the [last_error](https://github.com/dmlc/tvm/pull/3227/files/1791a7b5d00c2501683bc430e326c737e3c1c11c#diff-f1c1afaf885ccf923995f9bffd07bfdcR49) global variable in the runtime, because the session has access to the runtime's symbol table.  But if we now duplicate `last_error` in every loaded binary, the session needs to have access to the location of the `last_error` variable in the module of the function being run.

Mutinifni commented on src/runtime/micro/device/utvm_runtime.c:
```diff
@@ -0,0 +1,77 @@
+  num_active_allocs--;
+  if (num_active_allocs < 0) {
+    TVMAPISetLastError("free called with no active workspace allocations");
+    return -1;
```
`num_active_allocs` is -1 at this point, this might not be a consistent state for future allocations. This might matter if we choose to keep this semantics similar to `free` in C stdlib, because `free(NULL)` is considered a no-op.

weberlo replied:
Oh.  Good point.  Didn't think about resetting it for future task executions.

Mutinifni commented on src/runtime/micro/host_low_level_device.cc:
```diff
@@ -0,0 +1,99 @@
+  explicit HostLowLevelDevice(size_t num_bytes)
+    : size_(num_bytes) {
+    size_t size_in_pages = (num_bytes + kPageSize - 1) / kPageSize;
+    int mmap_prot = PROT_READ | PROT_WRITE | PROT_EXEC;
```
Note that the entire HostLowLevelDevice memory is executable. Do we want to change this to only be the code section?

weberlo replied:
I think it would definitely be helpful to apply the appropriate permissions to every section.  Personally, I'd like to add a `TODO` and implement it in a future PR, but if anyone has strong opinions on it, we can implement that in the current PR.

Mutinifni replied:
I think it's fine for now, and simplifies testing.

Mutinifni commented on src/runtime/micro/micro_session.cc:
```diff
@@ -0,0 +1,334 @@
+  std::string device_type = args[0];
+  if (device_type == "host") {
+    low_level_device_ = HostLowLevelDeviceCreate(kMemorySize);
+    SetInitBinaryPath(args[1]);
```
It might be helpful to document expected args contents.

u99127 commented:
I've been trying to play a bit with this , after doing a build with the branch I tried running 

$> python test_runtime_micro.py gives me a segfault with the following backtrace. 

Inspecting this in gdb after doing a build which has -g added to the flags,  I see the following crash with the backtrace as below. 

#0  tvm::runtime::MicroSectionAllocator::Free (this=0x0, offs=...) at /home/ramana/tvm-test/tvm/src/runtime/micro/micro_session.h:85
#1  0x00007fffacfe96e4 in tvm::runtime::MicroSession::FreeInSection (this=<optimised out>, type=type@entry=tvm::runtime::SectionKind::kHeap, ptr=...)
    at /home/ramana/tvm-test/tvm/src/runtime/micro/micro_session.cc:144
#2  0x00007fffacfde3d5 in tvm::runtime::MicroDeviceAPI::FreeDataSpace (this=<optimised out>, ctx=..., ptr=<optimised out>) at /home/ramana/tvm-test/tvm/src/runtime/micro/micro_device_api.cc:59
#3  0x00007fffacf88ec1 in tvm::runtime::NDArray::Internal::DefaultDeleter (ptr=0x124d3e0) at /home/ramana/tvm-test/tvm/src/runtime/ndarray.cc:62
#4  0x00007fffacf82060 in tvm::runtime::NDArray::Container::DecRef (this=<optimised out>) at /home/ramana/tvm-test/tvm/include/tvm/runtime/ndarray.h:326
#5  TVMArrayFree (handle=<optimised out>) at /home/ramana/tvm-test/tvm/src/runtime/ndarray.cc:220
#6  0x00007fffae64bdae in ffi_call_unix64 () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#7  0x00007fffae64b71f in ffi_call () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#8  0x00007fffae85f524 in _ctypes_callproc () from /home/ramana/tensorflow/v/lib/python3.6/lib-dynload/_ctypes.cpython-36m-x86_64-linux-gnu.so
#9  0x00007fffae85fb93 in ?? () from /home/ramana/tensorflow/v/lib/python3.6/lib-dynload/_ctypes.cpython-36m-x86_64-linux-gnu.so
#10 0x00000000005a730c in _PyObject_FastCallKeywords ()
#11 0x0000000000503073 in ?? ()
#12 0x0000000000506859 in _PyEval_EvalFrameDefault ()
#13 0x0000000000501945 in _PyFunction_FastCallDict ()
#14 0x0000000000591461 in ?? ()
#15 0x00000000005a337c in _PyObject_FastCallDict ()
#16 0x000000000061a2bd in ?? ()
#17 0x000000000055e895 in PyObject_CallFinalizerFromDealloc ()
#18 0x0000000000554567 in ?? ()
#19 0x0000000000504f98 in ?? ()
#20 0x0000000000502540 in ?? ()
#21 0x0000000000502f3d in ?? ()
#22 0x0000000000506859 in _PyEval_EvalFrameDefault ()
#23 0x0000000000504c28 in ?? ()
#24 0x0000000000506393 in PyEval_EvalCode ()
#25 0x0000000000634d52 in ?? ()
#26 0x0000000000634e0a in PyRun_FileExFlags ()
#27 0x00000000006385c8 in PyRun_SimpleFileExFlags ()
#28 0x000000000063915a in Py_Main ()
#29 0x00000000004a6f10 in main ()




tvm::runtime::MicroSectionAllocator::Free (this=0x0, offs=...) at /home/ramana/tvm-test/tvm/src/runtime/micro/micro_session.h:85
85	    CHECK(alloc_map_.find(ptr) != alloc_map_.end()) << "freed pointer was never allocated";
(gdb) bt
#0  tvm::runtime::MicroSectionAllocator::Free (this=0x0, offs=...) at /home/ramana/tvm-test/tvm/src/runtime/micro/micro_session.h:85
#1  0x00007fffacfe96e4 in tvm::runtime::MicroSession::FreeInSection (this=<optimised out>, type=type@entry=tvm::runtime::SectionKind::kHeap, ptr=...)
    at /home/ramana/tvm-test/tvm/src/runtime/micro/micro_session.cc:144
#2  0x00007fffacfde3d5 in tvm::runtime::MicroDeviceAPI::FreeDataSpace (this=<optimised out>, ctx=..., ptr=<optimised out>) at /home/ramana/tvm-test/tvm/src/runtime/micro/micro_device_api.cc:59
#3  0x00007fffacf88ec1 in tvm::runtime::NDArray::Internal::DefaultDeleter (ptr=0x124d3e0) at /home/ramana/tvm-test/tvm/src/runtime/ndarray.cc:62
#4  0x00007fffacf82060 in tvm::runtime::NDArray::Container::DecRef (this=<optimised out>) at /home/ramana/tvm-test/tvm/include/tvm/runtime/ndarray.h:326
#5  TVMArrayFree (handle=<optimised out>) at /home/ramana/tvm-test/tvm/src/runtime/ndarray.cc:220
#6  0x00007fffae64bdae in ffi_call_unix64 () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#7  0x00007fffae64b71f in ffi_call () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#8  0x00007fffae85f524 in _ctypes_callproc () from /home/ramana/tensorflow/v/lib/python3.6/lib-dynload/_ctypes.cpython-36m-x86_64-linux-gnu.so
#9  0x00007fffae85fb93 in ?? () from /home/ramana/tensorflow/v/lib/python3.6/lib-dynload/_ctypes.cpython-36m-x86_64-linux-gnu.so

This is on Ubuntu 18.04 . I'm happy to give any more details that you need.

weberlo commented:
@u99127 Oh.  That's embarassing.  I think I made a change and forgot to rebuild the C++ before I ran the Python test.  I've been looking into this, and it's a strange problem.  It has to do with [these lines](https://github.com/dmlc/tvm/pull/3227/files#diff-0c34c29829086585f73b79fc911d0cdcR90) that I added to destruct the allocators.

Take this code snippet, for example:
```python
with HOST_SESSION as sess:
    ctx = tvm.micro_dev(0)
    a = tvm.nd.array(..., ctx)

# Garbage collection of `a` happens here.
```
`EndSession` gets called when the `with` block ends, but the tensor `a` doesn't get freed until Python garbage collects them.  At that point, the allocators don't exist and can't be used to deallocate the tensor.

@tqchen What can we do to deal with this problem?  One possible fix is to change [FreeInSection](https://github.com/dmlc/tvm/pull/3227/files#diff-0c34c29829086585f73b79fc911d0cdcR123) to return immediately if the session isn't active, but this could still cause problems if we have two consecutive sessions:
```python
with HOST_SESSION as sess:
    ctx = tvm.micro_dev(0)
    a = tvm.nd.array(..., ctx)

with HOST_SESSION as sess:
    ctx = tvm.micro_dev(0)
    a = tvm.nd.array(..., ctx)
    # Garbage collection of the first `a` happens here.
```

tqchen commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,178 @@
+        self.binutil_prefix = binutil_prefix
+        self.port = port
+
+    def micro_build(self, func: relay.Function, params={}):
```
keep this function as global as it is not dependent on session.  We should minimize the number of functions, In this case. this function should simply be removed

Mutinifni replied:
Also change `params` default value from `{}` to `None`.

https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python/26320938

tqchen commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,178 @@
+        micro_mod = tvm.module.load(lib_obj_path, "micro_dev")
+        return micro_mod
+
+    def create_micro_lib(self, src_path, obj_path=None):
```
keep this function in global namespace

tqchen replied:
I will do ```tvm.micro.create_lib(src_path, "result.obj", cc="riscv64-unknown-elf-gcc")```

I agree we might need something in the session for binutils(for relocation etc), but I do view create_lib is something that everyone can use and do not necessarily have to be tied to session.

tqchen commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,178 @@
+        mod_src = c_mod.get_source()
+        with open(lib_src_path, "w") as f:
+            f.write(mod_src)
+        # Compile to object file.
```
remove this function, instead, directly support tvm.micro.create_lib(c_mod, "mylib.obj")

tqchen commented:
@weberlo Regarding the lifetime issue of Session. I was expecting this, but did not expect it to come so soon :) Here is a solution: keep a reference to shared_ptr<Session> in both allocated objects, Module, and PackedFunc.

More specifically:
- EnterSession will reset the Session::Global to a new Session 
- The data, function, and module keeps a shared_ptr to the session, and use the session through the shared_ptr. Note that during the creation of a Module, the shared_ptr is obtained by Session::Global()
- ExitSession is called, Session::Global shared_ptr is reset to nullptr.
- At this moment, the old data, function, and module still keep the reference to the old session shared_ptr, the session is not really "closed" until all these resources went out of the scope.
- We need to make sure to compare session ptr when during DeviceAPI::Copy, since you cannot copy a data from a stale session to a new one.

weberlo commented:
@tqchen This seems like a good solution.  One thing that still worries me is the situation where there are two concurrent sessions that both use the same underlying physical device.

This isn't a problem with the `HostLowLevelDevice`, because each instance of the class corresponds to a separate "physical device".  But this becomes a problem with the `OpenOCDLowLevelDevice`, because each instance of that class is connecting to a single physical device.  In that case, when there are still pointers to the first session, a second OpenOCD session cannot be created, because the socket address will still be in use by the first session.

tqchen commented:
In that case, we will need to make sure all the resources are out of scope, usually wrapping everything related in a function will help to resolve the issue.

u99127 commented:
It looks to me like $TVMHOME/python/setup.py is missing the 'nose' package as a dependency . Found by  trying to execute test_runtime_micro.py in a fresh environment.

weberlo commented:
@tqchen There's not a distinction between user and developer dependencies, is there?  `nose` makes sense as a developer dependency, because we'll only need it for tests, but maybe not as a user dependency.

MarisaKirisame commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,133 @@
+    src_path : str
+        path to source file
+
+    obj_path : str, optional
```
Optional[String]

weberlo replied:
It seems like most Python files already aren't following MyPy style.  Is there a push to change this?

weberlo replied:
@MarisaKirisame ^

MarisaKirisame commented on python/tvm/micro/cross_compile.py:
```diff
@@ -0,0 +1,63 @@
+    options: list
+        list of additional option strings
+
+    compile_cmd : str, optional
```
Optional[String]

MarisaKirisame commented on python/tvm/micro/cross_compile.py:
```diff
@@ -0,0 +1,63 @@
+    sources : list
+        list of source files to be compiled
+
+    options: list
```
List[Optional[String]]

MarisaKirisame commented on python/tvm/micro/cross_compile.py:
```diff
@@ -0,0 +1,63 @@
+
+    Parameters
+    ----------
+    output : str
```
String

MarisaKirisame commented on python/tvm/ndarray.py:
```diff
@@ -189,6 +189,22 @@ def ext_dev(dev_id=0):
+
+    Parameters
+    ----------
+    dev_id : int, optional
```
Optional[Int]

MarisaKirisame commented on src/codegen/codegen_c_host.h:
```diff
@@ -48,7 +48,8 @@ class CodeGenCHost final : public CodeGenC {
-  std::string module_name;
+  std::string module_name_;
+
   void PrintGetFuncFromBackend(std::string func_name, std::string packed_func_name);
```
const &

weberlo replied:
done

weberlo commented on python/tvm/contrib/binutil.py:
```diff
@@ -0,0 +1,260 @@
+    *(.sdata)
+  }
+}
+    """ % (text_addr, rodata_addr, data_addr, bss_addr)
```
This linker script generation feels a little bit ad hoc.  If anyone has ideas on a better way to do this, I'd be happy to hear them.  The only thing I can think of right now is to expose a linker configuration interface to the user.

I'm also perfectly happy punting this off to a future PR, when we've tested more hardware backends and figured out the pain points in the current interface.

tqchen commented on tests/python/unittest/test_runtime_micro.py:
```diff
@@ -173,7 +234,7 @@ def test_resnet_random():
                                        resnet_func.ret_type)
 
-    with micro.Session(DEVICE_TYPE, TOOLCHAIN_PREFIX):
+    with micro.create_session(DEVICE_TYPE, TOOLCHAIN_PREFIX):
```
Session is fine as it is a constructor

weberlo replied:
oh. I was just trying to mirror the graph runtime. I'll change it back.

weberlo commented on src/runtime/micro/micro_session.h:
```diff
@@ -0,0 +1,257 @@
+      CHECK(curr_session != nullptr) << "null global session";
+    }
+    return curr_session;
+  }
```
This is a major hack.  `Global` is being used as both a getter and setter, so they're both referencing the same static variable.  I tried pulling the variable out to be a class static variable, but that gave me "undefined symbol' errors at runtime.  I'm sure there's a better way to do this, but I'm not sure if I'm enough of a C++ guru to know how.

tqchen commented:
@weberlo please look into the ci error, will do a round of review today

tqchen commented:
Please be careful not to include generated files into the repo( I saw resnet.c etc)

weberlo commented:
@tqchen Yeah.  I'll remove it in a later commit.  Right now, I'm just using it to debug, because CI doesn't give me enough information to see what's going wrong.

weberlo commented:
@tqchen I've run the failing test on a Linux machine and on the ci_cpu docker image on my local machine, and I can't reproduce the failing test.  Is there any way I can get access to the CI machine itself to debug?

tqchen commented:
Have you used the exact same script? ```scripts/task_python_unittest.sh```. The problem might arise when you run everything together in a single process, but not when you run the single file

tqchen commented on include/tvm/runtime/micro/utvm_device_lib.h:
```diff
@@ -0,0 +1,71 @@
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
```
move to src/tvm/runtime/micro/device/utvm_device_lib.cc

Mainly because this header is used as "internal implementation", rather than a public header that a user can include

tqchen commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,131 @@
+        in a `with` block.
+
+        Parameters
+        ----------
```
move parameter documentation to the documentation of the class. See https://github.com/dmlc/tvm/blob/master/python/tvm/expr.py#L163

tqchen commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,131 @@
+        path to source file
+
+    obj_path : str, optional
+        path to generated object file (defaults to same directory as
```
I would recommend remove this default behavior and force the user to pass in a path. See our normal testcase https://github.com/dmlc/tvm/blob/master/tutorials/cross_compilation_and_rpc.py#L123

tqchen commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,131 @@
+        micro_device_dir = os.path.join(micro_dir, "..", "..", "..",
+                                        "src", "runtime", "micro", "device")
+        runtime_src_path = os.path.join(micro_device_dir, "utvm_runtime.c")
+        tmp_dir = util.tempdir()
```
NOTE: tmp_dir will get deleted when it goes out of scope. If you want to keep it alive, you need to keep it as a class member. This could be the reason why your test failed.

weberlo replied:
I don't think this is the problem.  The `_CreateSession` call below reads from `tmp_dir` and then `tmp_dir` is no longer needed.

tqchen commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,131 @@
+        self._enter()
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        pass
```
should reset the global scope?

tqchen commented on src/runtime/micro/device/utvm_runtime.h:
```diff
@@ -0,0 +1,98 @@
+ * certain backends such as OpenGL.
+ * \return nullptr when error is thrown, a valid ptr if success
+ */
+void* TVMBackendAllocWorkspace(int device_type,
```
not necessary as the function is declared in c backend api

tqchen commented on src/runtime/micro/device/utvm_runtime.c:
```diff
@@ -0,0 +1,87 @@
+UTVMTask task;
+
+// These pointers are patched at load time to point to the workspace section.
+char *utvm_workspace_begin = NULL;  // NOLINT(*)
```
char* utvm_workspace_begin

tqchen commented on src/runtime/micro/device/utvm_runtime.c:
```diff
@@ -0,0 +1,87 @@
+char *utvm_workspace_end = NULL;  // NOLINT(*)
+char *utvm_workspace_curr = NULL;  // NOLINT(*)
+// Keep track of how many active allocations there are on the workspace.
+size_t num_active_allocs = 0;
```
always prefix with utvm

tqchen commented on src/runtime/micro/device/utvm_runtime.c:
```diff
@@ -0,0 +1,87 @@
+UTVMTask task;
+
+// These pointers are patched at load time to point to the workspace section.
+char *utvm_workspace_begin = NULL;  // NOLINT(*)
```
consider making the global state a struct. UTVMContext

weberlo replied:
I'm not sure if we can do this in a clean way, because we rely on the symbols being exposed at the global level to be able to read from and write to them.  If it's some `utvm_context` struct being exposed, then we would need to manually calculate the offset of `workspace_begin` to patch that pointer.  I'm hesitant to add any more fragility, because we still don't have a great solution for maintaining a correspondence between struct padding on the device side and the host side.  Let me know what you think.

tqchen commented on src/runtime/micro/device/utvm_runtime.h:
```diff
@@ -0,0 +1,98 @@
+  TVMValue* values;
+  /*! \brief Array of type codes for each value */
+  int* type_codes;
+  /*! \brief Number of arguments */
```
can we inline UTVMArgs's field into UTVMTask?

tqchen commented on src/runtime/micro/host_low_level_device.h:
```diff
@@ -0,0 +1,81 @@
+namespace runtime {
+/*!
+ * \brief emulated low-level device on host machine
+ */
```
We do not need to expose HostLowlevelDevice into a header. So to minimize the number of .h files. Consider directly move it into the implementation (host_low_level_device.cc).

Move the declaration of HostLowLevelDeviceCreate into low_level_device.h

tqchen commented on src/runtime/micro/micro_common.cc:
```diff
@@ -0,0 +1,184 @@
+namespace tvm {
+namespace runtime {
+
+DevBaseOffset DevAddr::operator-(DevBaseAddr base) const {
```
consider inline those implementations into DevAddr as inline functions. Because they are short, and should be inlined to the caller(rather than become its own function)

tqchen commented on src/runtime/micro/micro_device_api.cc:
```diff
@@ -0,0 +1,173 @@
+          << " != " << to_space->session << ")";
+      CHECK(ctx_from.device_id == ctx_to.device_id)
+        << "can only copy between the same micro device";
+      std::shared_ptr<MicroSession> session = from_space->session;
```
use reference instead of deep copy, this will increase the reference counter of the session

tqchen commented on src/runtime/micro/micro_device_api.cc:
```diff
@@ -0,0 +1,173 @@
+      // Reading from the device.
+
+      MicroDevSpace* from_space = static_cast<MicroDevSpace*>(const_cast<void*>(from));
+      std::shared_ptr<MicroSession> session = from_space->session;
```
use reference instead of deep copy, this will increase the reference counter of the session, can inline into the next line

tqchen commented on src/runtime/micro/micro_module.cc:
```diff
@@ -0,0 +1,150 @@
+    const std::shared_ptr<ModuleNode>& sptr_to_self) {
+  DevBaseOffset func_offset = symbol_map()[name];
+  MicroWrappedFunc f(this, this->session_, name, func_offset);
+  return PackFuncVoidAddr(f, std::vector<TVMType>());
```
You do not need to use PackFunVoidAddr, this function packs the arguments into addresses that are needed by CUDA calling convention. You can directly create a PackedFunc that takes args and rv

tqchen commented on src/runtime/micro/micro_section_allocator.h:
```diff
@@ -0,0 +1,122 @@
+    : start_offset_(region.start),
+      size_(0),
+      capacity_(region.size) {
+      CHECK(start_offset_.value() % 8 == 0) << "micro section not aligned to 8 bytes";
```
CHECK_EQ

tqchen commented on src/runtime/micro/micro_session.h:
```diff
@@ -0,0 +1,257 @@
+   * \brief get MicroSession global singleton
+   * \return pointer to the micro session global singleton
+   */
+  static std::shared_ptr<MicroSession> Global(
```
Checkout https://github.com/dmlc/tvm/blob/master/include/tvm/build_module.h#L110

and code below https://github.com/dmlc/tvm/blob/master/src/codegen/build_module.cc#L237

tqchen commented:
@weberlo please see my comments about the potential reason for CI error https://github.com/dmlc/tvm/pull/3227#discussion_r300897448

tqchen commented:
@MarisaKirisame @liangfu @Mutinifni @u99127 please take around round of review

tqchen commented:
please rebase against the latest master as HalideIR was removed in the latest merge. @MarisaKirisame @u99127 please help do another round of review.

tqchen commented on src/runtime/micro/device/utvm_device_lib.h:
```diff
@@ -0,0 +1,55 @@
+
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file utvm_device_lib.h
```
utvm_device_dylib_redirect.cc

Add comments to make the purpose of this file clear: This is a library that gets linked to each utvm library. We redirect each library call into a pre-defined global function pointer, and we patch the correct address of the library into those pointers when we load the library.

tqchen commented on src/runtime/micro/micro_session.h:
```diff
@@ -0,0 +1,244 @@
+   * \brief returns low-level device pointer
+   * \note assumes low-level device has been initialized
+   */
+  const std::shared_ptr<LowLevelDevice> low_level_device() const {
```
return const reference, so you don't have to copy the pointer everytime

tqchen commented on src/runtime/micro/micro_session.h:
```diff
@@ -0,0 +1,244 @@
+    return low_level_device_;
+  }
+
+  SymbolMap& init_symbol_map() {
```
is this only used by member function? if so, move it to private. It is a bit uncommon to expose a non-const reference member

weberlo replied:
It's used [here](https://github.com/dmlc/tvm/pull/3227/files#diff-a017c34eeae79dcd77b0b23e6c3fd009R97), but now that I think about it, `PatchImplHole` should be in `MicroSession` and not in `MicroModule`.  I'll move it, then make `init_symbol_map` private.

tqchen commented on src/runtime/micro/micro_session.h:
```diff
@@ -0,0 +1,244 @@
+ * only deallocate the session once there are no more references to it.
+ */
+struct MicroDevSpace {
+  void* data;
```
document each field.

tqchen commented:
Some additional comment about API, right now we use relay_micro_build in test to encapsulate a lot of things, while it is good, ideally, we want to remove as many abstractions as possible and reuse the APIs in a compositional way. Here is one possible mock up API.

```python
    with tvm.build_config(disable_vectorize=True):
        graph, c_mod, params = relay.build(func, target="c", params=params)
    c_mod.export_library("mylib.obj", tvm.micro.cross_compiler(prefix=""))
    micro_mod = tvm.module.load("mylib.obj", "micro_dev") 
    ctx = tvm.micro_dev(0)
    mod = graph_runtime.create(graph, micro_mod, ctx)
    mod.set_input(**params)

```

weberlo commented:
@tqchen So the main change you're proposing is that we expose `tvm.micro.cross_compiler` as a function that can be used with `export_library`?  Should we also keep `create_micro_lib`, in case users want to supply their own source files and not be restricted to source files generated by TVM?

tqchen commented:
tvm.micro.cross_compiler returns a functor that can take a list of source files, see also https://github.com/dmlc/tvm/blob/master/python/tvm/contrib/cc.py#L60

u99127 commented on cmake/config.cmake:
```diff
@@ -62,6 +62,9 @@ set(USE_VULKAN OFF)
 # Whether enable OpenGL runtime
 set(USE_OPENGL OFF)
 
+# Whether enable Micro runtime
```
Would microcontroller be a better name for the runtime rather than micro runtime?

weberlo replied:
I think in that comment, it's supposed to be short for "MicroTVM", but we can just write it out as "MicroTVM" to make it clearer.

weberlo commented:
@u99127 Yep.  That's correct.

u99127 commented on python/tvm/micro/base.py:
```diff
@@ -0,0 +1,190 @@
+    options += ["-fno-stack-protector"]
+    # TODO(weberlo): This option cannot be used on 32-bit machines. Make this
+    # compilation pipeline compatible with 32-bit.
+    options += ["-mcmodel=large"]
```
This option should be host based . Usually the '-m' prefixed options are machine specific and hard coded. While -mcmodel=large would likely work on other hosts, we should aim to keep this largely host clean . 

Maybe a check that this is an x86_64-linux host might be appropriate here especially as you are looking to create a large model .so ?

u99127 commented on src/runtime/micro/micro_session.cc:
```diff
@@ -0,0 +1,371 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file micro_session.cc
+ * \brief session to manage multiple micro modules
```
Is this envisaging the presence of multiple micro modules connected to the same host and the capability of running parallel sessions on all of them ? Some more detail here would be good.

tqchen commented on python/tvm/contrib/binutil.py:
```diff
@@ -0,0 +1,260 @@
+    toolchain_prefix : str
+        prefix for binary names in target compiler toolchain
+
+    Return
```
Return-> Returns

weberlo commented on tests/python/contrib/test_binutil.py:
```diff
@@ -0,0 +1,145 @@
+    tmp_obj = tmp_dir.relpath("obj.o")
+    with open(tmp_source, "w") as f:
+        f.write(prog)
+    p1 = subprocess.Popen(["gcc", "-c", tmp_source, "-o", tmp_obj],
```
Good point.  I also realize we can use `cc.create_shared` here instead of using a subprocess.

weberlo commented on python/tvm/contrib/binutil.py:
```diff
@@ -0,0 +1,260 @@
+        ".text": [".text"],
+        ".rodata": [".rodata"],
+        ".data": [".data"],
+        ".bss": [".bss", ".sbss", ".sdata"],
```
Would it be fine if it's mapped with `.data`?  Or should it be its own section?

weberlo replied:
Nevermind.  Just saw your comment below.

tqchen commented:
Thanks @weberlo @u99127 @MarisaKirisame ! This PR is now merged

apivovarov commented:
@weberlo, @tqchen What you think if we rename `compile_cmd` param back to `cc` in `lib.export()`?

weberlo commented:
@apivovarov Which part of the PR are you referring to?

apivovarov commented:
@weberlo  https://discuss.tvm.ai/t/lib-export-library-cc-param-was-renamed-to-compile-cmd/3644

weberlo commented:
@apivovarov Just made a PR to remedy this (#3746).  Sorry for the trouble.

apivovarov commented:
Thank you!
